File: java/src/main/java/Evaluate_Division_399/Solution.java
package Evaluate_Division_399;

import java.util.*;

public class Solution {
  public double[] calcEquation(String[][] equations, double[] values, String[][] queries) {

    double[] ans = new double[queries.length];

    // Edge cases
    if (equations == null || values == null || equations.length != values.length) return null;

    HashMap<String, Vertex> map = new HashMap<>();

    // Populate the map of vertices and edges
    for (int i = 0; i < equations.length; i++) {
      String aName = equations[i][0];
      String bName = equations[i][1];
      Vertex a;
      Vertex b;

      // Get vertex from map or create a new one
      if (map.containsKey(aName)) {
        a = map.get(aName);
      } else {
        a = new Vertex(aName);
        map.put(aName, a);
      }

      if (map.containsKey(bName)) {
        b = map.get(bName);
      } else {
        b = new Vertex(bName);
        map.put(bName, b);
      }

      // Get the value
      double w = values[i];

      // Add edges with weight to the vertices
      a.addConnection(b, w);
      b.addConnection(a, 1/w);
    }

    // Go through queries
    for (int i = 0; i < queries.length; i++) {
      String aName = queries[i][0];
      String bName = queries[i][1];

      if (!map.containsKey(aName) || !map.containsKey(bName)) {
        ans[i] = -1.00;
      } else {
        ans[i] = getPath(map.get(aName), map.get(bName), 1.00);
      }
    }

    return ans;
  }

  private double getPath(Vertex a, Vertex b, double weight) {
    Set<Vertex> visited = new HashSet<>();
    Queue<VertexWeight> q = new LinkedList<>();

    q.add(new VertexWeight(a, 1));
    VertexWeight curr = q.peek();

    while (!q.isEmpty() && curr.vertex != b) {
      curr = q.poll();
      visited.add(curr.vertex);

      for (Edge edge : curr.vertex.connections) {
        Vertex v = edge.end;
        Double p = edge.weight;

        // Handle only not visited vertices
        if (!visited.contains(v)) {
          q.add(new VertexWeight(v, p * curr.weight));
        }
      }

    }
    return curr.vertex == b ? curr.weight : -1 ;
  }

  private class VertexWeight {
    Vertex vertex;
    double weight;

    VertexWeight(Vertex vertex, double weight) {
      this.vertex = vertex;
      this.weight = weight;
    }
  }

  private class Vertex {
    String name;
    HashSet<Edge> connections;

    Vertex(String name) {
      this.name = name;
      connections = new HashSet<>();
    }

    void addConnection(Vertex v, double weight) {
      Edge e = new Edge(this, v, weight);
      connections.add(e);
    }

    public String toString() {
      return this.name;
    }
  }

  private class Edge {
    Vertex start;
    Vertex end;
    double weight;

    Edge(Vertex start, Vertex end, double weight) {
      this.start = start;
      this.end = end;
      this.weight = weight;
    }

    public String toString() {
      return start + "->" + end + " (" + weight + ")";
    }
  }
}

File: java/src/main/java/Evaluate_Division_399/solution.py
class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        # equations: [a/b], [b/c]
        # values: [2.0,3.0]
        # queries: [a/c],[b/a],[a/e],[a/a],[x/x]

        graph = defaultdict(defaultdict)

        def dfs(curr, target, p, visited):
            visited.add(curr)
            ans = -1.0
            neighbors = graph[curr]
            if target in neighbors:
                ans = p * neighbors[target]
            else:
                for neighbor, val in neighbors.items():
                    if neighbor in visited:
                        continue
                    ans = dfs(neighbor, target, p * val, visited)
                    if ans != -1.0:
                        break
            visited.remove(curr)
            return ans

        # Build graph from equations and values
        for (a, b), val in zip(equations, values):
            graph[a][b] = val
            graph[b][a] = 1 / val

        # Evaluate each query
        results = []
        for a, b in queries:
            if a not in graph or b not in graph:
                results.append(-1.0)
            elif a == b:
                results.append(1.0)
            else:
                visited = set()
                res = dfs(a, b, 1, visited)
                results.append(res)
        
        return results
        